pkgs/development/compilers/dotnet/default.nix

```diff
@@ -30,21 +27,57 @@ let
     "i686-windows" = "win-x86";
   };
 
-  # Convert a "stdenv.hostPlatform.system" to a dotnet RID
-  systemToDotnetRid = system: runtimeIdentifierMap.${system} or (throw "unsupported platform ${system}");
-in
-{
-  inherit systemToDotnetRid;
-
   combinePackages = attrs: callPackage (import ./combine-packages.nix attrs) {};
 
-  dotnet_8 = recurseIntoAttrs (callPackage ./8 { bootstrapSdk = dotnet_8_0_102.sdk_8_0; });
-  dotnet_9 = recurseIntoAttrs (callPackage ./9 {});
-} // lib.optionalAttrs config.allowAliases {
-  # EOL
-  sdk_2_1 = throw "Dotnet SDK 2.1 is EOL, please use 6.0 (LTS) or 7.0 (Current)";
-  sdk_2_2 = throw "Dotnet SDK 2.2 is EOL, please use 6.0 (LTS) or 7.0 (Current)";
-  sdk_3_0 = throw "Dotnet SDK 3.0 is EOL, please use 6.0 (LTS) or 7.0 (Current)";
-  sdk_3_1 = throw "Dotnet SDK 3.1 is EOL, please use 6.0 (LTS) or 7.0 (Current)";
-  sdk_5_0 = throw "Dotnet SDK 5.0 is EOL, please use 6.0 (LTS) or 7.0 (Current)";
-} // dotnet_6_0 // dotnet_7_0 // dotnet_8_0 // dotnet_9_0
+  makeDotnet = args:
+    let self =
+      lib.makeOverridable ({ sdk, runtime, aspnetcore, ...}@pkgs:
+        recurseIntoAttrs (
+          pkgs // {
+            buildDotnetModule = buildDotnetModule.override { dotnet = self; };
+            buildDotnetGlobalTool = buildDotnetGlobalTool.override { dotnet = self; };
+            nuget-to-nix = nuget-to-nix.override { dotnet = self; };
+
+            withExtraSDKs = sdks: self.override {
+              sdk = combinePackages (lib.singleton sdk ++ sdks);
+            };
+          })) args;
+    in self;
+
+  pkgs = rec {
+    # Convert a "stdenv.hostPlatform.system" to a dotnet RID
+    systemToDotnetRid = system: runtimeIdentifierMap.${system} or (throw "unsupported platform ${system}");
+
+    inherit combinePackages;
+
+    ## Files in versions/ are generated automatically by update.sh ##
+    dotnet_6-bin = makeDotnet (import ./versions/6.0.nix buildAttrs);
+    dotnet_7-bin = makeDotnet (import ./versions/7.0.nix buildAttrs);
+    dotnet_8-bin = makeDotnet (import ./versions/8.0.nix buildAttrs);
+    dotnet_8_0_102-bin = makeDotnet (import ./versions/8.0.102.nix buildAttrs);
+    dotnet_9-bin = makeDotnet (import ./versions/9.0.nix buildAttrs);
+
+    # We don't have source-builds of 6/7
+    dotnet_6 = dotnet_6-bin;
+    dotnet_7 = dotnet_7-bin;
+    dotnet_8 = makeDotnet (callPackage ./8 { bootstrapSdk = dotnet_8_0_102-bin.sdk; });
+    dotnet_9 = makeDotnet (callPackage ./9 {});
```

Comment:
I'm a bit on the fence about this because of .NET 8, we'd be providing an intuitive default (`dotnet_8`) which is actually just a maintenance version instead of a more featureful one.

Maybe for the time being, we suffix the source build ones and not the binary ones? Or prefix both

Reply:
I think we should probably discourage use of the binary packages, especially due to #169427.

Reply:
I mean, according to what omajid said:

> The LICENSE file for the linux SDK tarballs, as far as I am aware, is not the proprietary EULA.

We don't need to worry about that since we only handle mac and linux, if we handled windows then that'd be a concern.

I just downloaded every runtime, SDK and ASP.NET Core runtime tarball from MS for the latest versions of .NET 6, 7 and 8 and am reading through the licenses and I don't see any proprietary ones so far.

Reply:
Ah, okay, maybe we should close that then.

Reply:
I've left a comment there noting this out, I'll leave it up to the people already in there to decide whether to keep it open or not, but thanks for bringing it up

Reply:
I still think it would be a good idea to use source-built packages as much as possible.  It should allow us to eventually make the dotnet ecosystem more nix-friendly.  For example, dropping the elf patching stuff for system libs.  We could also add support for locating sdks, runtimes, and nuget packages from multiple locations in /nix/store.

I do share your concern about the out of date release though.  There's also a question over cross-compilation, because the source built SDKs only have packages for the target platform, and I haven't been able to build them in cross yet.

Reply:
> I still think it would be a good idea to use source-built packages as much as possible. It should allow us to eventually make the dotnet ecosystem more nix-friendly.

Yes, I am in favor of that for .NET 9, however for .NET 8 I think it'd be better for the user to have a more featureful SDK than one that is source-built when they have no visible benefit (since we already do patching for the binary one, I don't think they'll see much difference).

Reply:
In case it's not obvious (it wasn't to me at first), please note that 8.0.1xx and 8.0.2xx are both maintained in parallel. Both are getting bugfixes.

In fact, 8.0.1xx will be maintained for the entire duration of .NET 8's lifecycle, while 8.0.2xx will be EOL'ed by upstream when the associated Visual Studio version goes EOL. There would be a 8.0.3xx (and possibly 8.0.4xx, and so on) to replace it, though.

Reply:
I believe that the packages without a minor version (e.g. `dotnet_8`) should point to the latest SDK (e.g. 8.0.2xx or 8.0.4xx, whichever is latest).

I'm aware that 8.0.1xx will be maintained throughout the lifecycle but I think users will prefer to have the latest version as opposed to just the initial one but with bugfixes.

Reply:
> I think users will prefer to have the latest version as opposed to just the initial one but with bugfixes.

We could make the latest one usable via e.g. `dotnet-sdk_8`, but still use the source-built one inside nixpkgs wherever possible.

There's lots of precedent for `-bin` packages, but nothing like `-src` that I can see.  I'm not sure if there's anything where the latest version is only available in `-bin`.